libevm/VM.cpp

```diff
@@ -621,6 +621,66 @@ void VM::interpretCases()
 		}
 		NEXT
 
+#ifdef EVM_USE_BITSHIFT
+		CASE(SHL)
+		{
+			ON_OP();
+			updateIOGas();
+
+			if (m_SP[0] >= 256)
+				m_SPP[0] = 0;
+			else
+				m_SPP[0] = m_SP[1] << uint64_t(m_SP[0]);
+		}
+		NEXT
+
+		CASE(SHR)
+		{
+			ON_OP();
+			updateIOGas();
+
+			if (m_SP[0] >= 256)
+				m_SPP[0] = 0;
+			else
+				m_SPP[0] = m_SP[1] >> uint64_t(m_SP[0]);
+		}
+		NEXT
+
+		CASE(SAR)
+		{
+			ON_OP();
+			updateIOGas();
+
+			static u256 const hibit = u256(1) << 255;
```

Comment:
I think I'd prefer a slower, but definitely correct implementation for this case to ensure everything is working properly. Once multiple clients agree on tests, it can be optimised for speed.

Reply:
Do you think the code I have now is mathematically incorrect?  Or are you worried about bugs in boost's library?